File: Haskell/parser_combinator/parse_csv.hs
import Text.ParserCombinators.Parsec

csvFile = endBy line eol
line = sepBy cell (char ',')
cell = quotedCell <|> many (noneOf ",\n")
eol = char '\n'
quotedCell = do 
	char '"'
	content <- many quotedChar
	char '"'
	return content

quotedChar = noneOf "\"" <|> try (string "\"\"" >> return '"')

main = do
	cs <- getContents
	let res = parse csvFile "" cs

	case res of
		Left err -> print err
		Right x -> putStrLn $ show x

File: blog/20110103/csv_rparsec/parse_csv.rb

require 'rubygems'
require 'rparsec'

include RParsec::Parsers

eol = string "\r\n"
quotedChar = not_char('"') | string('""') >> value('"'[0])
quotedCell = char('"') >> quotedChar.many.bind {|s| value(s.pack("c*"))} << char('"')
cell = quotedCell | regexp(/[^,\r\n]*/)
line = cell.separated(char ',')
csvFile = (line << eol).many

cs = $stdin.readlines.join
res = csvFile.parse cs

p res
puts res
